src/recompiler.ts

```diff
@@ -0,0 +1,241 @@
+            changed.add(name);
+        }
+    }
+
+    return changed;
+};
+
+// Compile changed contracts
+const compileChangedContracts = async (
+    hre: HardhatRuntimeEnvironment,
+    newCacheEntry: Cache,
+    changed: Set<string>
+): Promise<void> => {
+    for (const changedContract of changed) {
+        const entry = newCacheEntry[changedContract];
+        const compileArguments: TaskArguments = {
+            paths: [changedContract],
+            disableHintValidation: entry?.disableHintValidation,
+            accountContract: entry?.accountContract,
+            carioPath: entry?.cairoPath
+        };
+
+        await starknetCompileAction(compileArguments, hre);
+    }
+};
+
+// Updated set with changed and new contracts
+const updateSet = async (
+    paths: ProjectPathsConfig,
+    cache: Cache,
+    newCacheEntry: Cache,
+    changed: Set<string>
+): Promise<Set<string>> => {
+    for (const contractName in newCacheEntry) {
+        // Add new contracts that are not in cache before
+        if (!cache[contractName]) {
+            changed.add(contractName);
+            continue;
+        }
+
+        // Add contracts that contiain a change in content
+        if (newCacheEntry[contractName].contentHash !== cache[contractName].contentHash) {
+            changed.add(contractName);
+        }
+    }
+
+    // Remove deleted sources from old cache by overwriting it
+    const { cache: cacheDirName } = paths;
+    const cacheFile = path.join(cacheDirName, CACHE_FILE_NAME);
+    // Save updated cache
+    await fsPromises.writeFile(cacheFile, JSON.stringify(newCacheEntry, null, " "));
+
+    return changed;
+};
+
+// Handles cache on Starknet cli calls
+export const handleCache = async (hre: HardhatRuntimeEnvironment) => {
+    // If recompile is not enabled, do nothing
+    if (!hre.userConfig?.starknet?.recompile) return;
+
+    // Get cache directory from hre.config
+    const { cache: cacheDirName } = hre.config.paths;
+
+    try {
+        const oldCache = await upsertFile(cacheDirName);
+        const newCacheEntry = await getContractHash(hre.config.paths);
+        const changedContracts = await checkArtifacts(hre.config.paths, newCacheEntry);
+        const updatedSet = await updateSet(hre.config.paths, oldCache, newCacheEntry, changedContracts);
+        await compileChangedContracts(hre, newCacheEntry, updatedSet);
+    } catch (error) {
+        // If there is an error, do not recompile
+        console.error(error);
+        process.exit(1);
+    }
+};
+
+// Updates cache with new contract and artifacts
+export const updateCache = async (
+    file: string,
+    output: string,
+    abi: string,
+    hre: HardhatRuntimeEnvironment,
+    args: TaskArguments,
+    cairoPath?: string
+): Promise<void> => {
+    const { cache: cacheDirName } = hre.config.paths;
+    const oldCache = await upsertFile(cacheDirName);
+    const newCacheEntry = await getCacheEntry(file, output, abi, args, cairoPath);
```

Comment:
We keep those arguments in the cache so that the next time recompilation is performed, all the arguments will be used. But currently, we don't recompile for a change in a `cairoPath` although we would update it if the user manually compiles with the updated `cairoPath`.